src/transform/src/canonicalize_mfp.rs

```diff
@@ -272,37 +276,69 @@ impl CanonicalizeMfp {
+        // the possible residuals for every literal constraint row, and check that all sets are
+        // equal. Example: The user wrote
+        // `WHERE ((a=1 AND b=1) OR (a=2 AND b=2)) AND (c OR (d AND e))`.
+        // The DNF of this is
+        // `(a=1 AND b=1 AND c) OR (a=1 AND b=1 AND d AND e) OR (a=2 AND b=2 AND c) OR (a=2 AND b=2 AND d AND e)`.
+        // Then `constraints_to_residual_sets` will be:
+        // [
+        //   [`a=1`, `b=1`]  ->  {[`c`], [`d`, `e`]},
+        //   [`a=2`, `b=2`]  ->  {[`c`], [`d`, `e`]}
+        // ]
+        // After removing the literal constraints we have
+        // `c OR (d AND e)`
+        let mut constraints_to_residual_sets = BTreeMap::new();
+        or_args.iter().for_each(|or_arg| {
+            let and_args = or_arg.and_or_args(VariadicFunc::And);
+            let (mut constraints, mut residual): (Vec<_>, Vec<_>) =
+                and_args.iter().partition(|and_arg| {
+                    key.iter()
                         .any(|key_field| matches!(and_arg.expr_eq_literal(key_field), Some(..)))
                 });
-                MirScalarExpr::CallVariadic {
-                    func: VariadicFunc::And,
-                    exprs: and_args,
-                }
-            })
-            .collect::<Vec<_>>();
-        assert!(or_args_residual.len() >= 1); // We already checked `or_args.len() == 0` above
-        let or_args_residual_first = or_args_residual.get(0).unwrap();
-        if or_args_residual
+            // In every or_arg there has to be some literal constraints, otherwise
+            // `detect_literal_constraints` would have returned None.
+            assert!(constraints.len() >= 1);
+            // `remove_impossible_or_args` made sure that inside each or_arg, each
+            // expression can be literal constrained only once. So if we find one of the
+            // key fields being literal constrained, then it's definitely that literal
+            // constraint that detect_literal_constraints based one of its return values on.
+            //
+            // This is important, because without `remove_impossible_or_args`, we might
+            // have the situation here that or_arg would be something like
+            // `a = 5 AND a = 8`, of which `detect_literal_constraints` found only the `a = 5`,
+            // but here we would remove both the `a = 5` and the `a = 8`.
+            constraints.sort();
+            residual.sort();
+            let entry = constraints_to_residual_sets
+                .entry(clone_elements(&constraints))
+                .or_insert_with(BTreeSet::new);
+            entry.insert(clone_elements(&residual));
+        });
+        let residual_sets = constraints_to_residual_sets
             .iter()
-            .all(|or_arg| or_arg == or_args_residual_first)
-        {
+            .map(|(_constraints, residual_set)| residual_set)
```

Comment:
Why is it necessary to create a BTreeMap when the keys of the BTreeMap are not used? Can you instead have a HashSet of residuals, and then check that the residuals are all equal by seeing if the set only has one item?

Reply:
The BTreeMap is needed because I'm accessing the keys in an arbitrary order. So it's not that first I completely build up the set that is mapped to the first key, and then I completely build up the set that is mapped to the second key, and so on. But instead sometimes I add an element to the set that's mapped to the first key, and then to the set that's mapped to the second key, and then maybe again to the set that's mapped to the first key.

Reply:
Makes sense. Though you could collect it into a `HashSet` instead and then check that the collected `HashSet` has length 1.